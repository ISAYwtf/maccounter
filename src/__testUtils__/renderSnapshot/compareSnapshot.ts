import { render } from '@testing-library/react'
import React, { ReactElement } from 'react'

export const compareSnapshot = (component: ReactElement, wrapper?: React.FC) => {
    const { asFragment } = render(component, { wrapper })
    expect(asFragment()).toMatchSnapshot()
}
