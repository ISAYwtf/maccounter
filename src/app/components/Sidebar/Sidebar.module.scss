.sidebar {
  position: sticky;
  top: 100px;
  padding: 20px;
  height: max-content;
  display: flex;
  flex-direction: column;
  transition: .2s;
  user-select: none;
  z-index: 999;
}

.prompt {
  position: absolute;
  transform-origin: left center;
  opacity: 0;
  pointer-events: none;

  background-color: var(--color-bg-component);
  padding: 10px 10px 10px 20px;
  border-radius: var(--size-border-rad-component);
  margin-left: -7px;
  overflow: hidden;
  transition: 0.2s;
  z-index: 0;
}

.sidebarItem {
  position: relative;
  display: flex;
  align-items: center;

  color: var(--color-disable);
  transition: .2s;
  cursor: pointer;

  --svg-color-fill: var(--color-bg-component);
  --svg-color-stroke: var(--color-disable);

  &:not(:last-child) {
    margin-bottom: 50px;
  }

  svg {
    width: 24px;
    height: 24px;
    transition: 0.08s;
    position: relative;
    fill: var(--svg-color-fill);
    z-index: 1;

    path {
      transition: 0.08s;
      stroke: var(--svg-color-stroke);
    }
  }
}

.sidebarProfile {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  color: var(--color-disable);
  border: 1px solid var(--color-disable);
  border-radius: 50%;
  font-size: var(--size-font-side);
  transition: .2s;
  position: relative;
  z-index: 1;
}

/* При наведении */

.sidebarItem:hover {

  .sidebarProfile {
    background-color: var(--color-main);
    border-color: var(--color-main);
  }

  svg {
    --svg-color-fill: var(--color-main);

    path {
      --svg-color-stroke: var(--color-bg-component);
    }
  }
}

/* При активной вкладке */
.active {
  .sidebarProfile {
    background-color: var(--color-disable);
    color: var(--color-bg-component);
  }

  svg {
    --svg-color-fill: var(--color-disable);

    path {
      --svg-color-stroke: var(--color-bg-component);
    }
  }
}

/* При наведении и ожидании 0.5с */

.promptActive.sidebarItem {
  position: relative;

  &.sidebarItem {
    svg {
      --svg-color-fill: var(--color-main);

      path {
        --svg-color-stroke: var(--color-bg-component);
      }
    }

    .sidebarProfile {
      background-color: var(--color-main);
      border-color: var(--color-bg-component);
    }
  }

  .prompt {
    transform: scaleX(1);

    opacity: 1;

    background-color: var(--color-main);
    padding: 10px 10px 10px 54px;
    color: var(--color-bg-component);
    z-index: 0;
  }
}

/* При двойном щелчке */

.opened {
  .sidebarItem {
    overflow-x: hidden;

    svg {

      path {
        --svg-color-stroke: var(--color-txt);
        transition: 0.2s;
      }
    }

    .sidebarProfile {
      border-color: var(--color-txt);
      color: var(--color-txt);
    }

    .prompt {
      //position: relative;
      pointer-events: unset;
      opacity: 1;
      transform: scaleX(1);
      color: var(--color-txt);
      margin-left: 20px;
      padding-right: 0;
    }

    &:hover {
      .sidebarProfile {
        background: none;
        border-color: var(--color-main);
        color: var(--color-main);
      }

      svg {
        --svg-color-fill: var(--color-bg-component);

        path {
          --svg-color-stroke: var(--color-main);
        }
      }

      .prompt {
        color: var(--color-main);
      }
    }
  }
}
