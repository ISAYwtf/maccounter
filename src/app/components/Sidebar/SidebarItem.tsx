import { BasePropsWithoutChild } from '@localTypes/BaseProps'
import React, { MouseEventHandler, useState } from 'react'
import { NavLink } from 'react-router-dom'
import clsx from 'clsx'
// eslint-disable-next-line css-modules/no-unused-class
import styles from './Sidebar.module.scss'

interface SidebarItemProps extends BasePropsWithoutChild {
    title: string,
    Ico: React.FC | undefined,
    link: string,
    isSidebarOpened: boolean,
    isLinked?: boolean,
}

const SidebarItem: React.FC<SidebarItemProps> = (({
    title,
    Ico,
    link,
    isSidebarOpened,
    className,
    isLinked,
    onClick,
}) => {
    const [isActive, setIsActive] = useState(false)
    const DELAY = 1000
    let timerForShow: ReturnType<typeof setTimeout>

    const hoverSidebar = () => {
        if (isSidebarOpened) { return }
        timerForShow = setTimeout(() => {
            setIsActive(true)
        }, DELAY)
    }

    const leaveSidebar = () => {
        clearTimeout(timerForShow)
        setIsActive(false)
    }

    const clickHandler: MouseEventHandler = (e) => {
        if (!isLinked) { e.preventDefault() }
        if (onClick) {
            leaveSidebar()
            onClick(e)
        }
    }

    return (
        <NavLink
            to={link}
            activeClassName={styles.active}
            role="link"
            className={clsx(
                styles.sidebarItem,
                { [styles.promptActive]: isActive },
                className,
            )}
            onMouseEnter={hoverSidebar}
            onMouseLeave={leaveSidebar}
            onClick={clickHandler}
        >
            {Ico ? <Ico /> : <span className={styles.sidebarProfile}>IA</span>}
            <div className={styles.prompt}>
                <p>{title}</p>
            </div>
        </NavLink>
    )
})

SidebarItem.defaultProps = { isLinked: true }

export default SidebarItem
