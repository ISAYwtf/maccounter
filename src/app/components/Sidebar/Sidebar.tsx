import { ReactComponent as accountancySvg } from '@icons/accountancy.svg'
import { ReactComponent as addSvg } from '@icons/add.svg'
import { ReactComponent as homeSvg } from '@icons/home.svg'
import { ReactComponent as settingsSvg } from '@icons/settings.svg'
import { ReactComponent as statisticsSvg } from '@icons/statistics.svg'
import React, { useCallback, useState } from 'react'
import clsx from 'clsx'
import SidebarItem from '@components/Sidebar/SidebarItem'
import { ReactComponent as expandSvg } from '@icons/expand.svg'
import { useSelector, useDispatch } from '@store/hooks'
import { setModalIsActive } from '@store/app-store/appReducer'
import { nameParser } from '@utils/stringParser'

// eslint-disable-next-line css-modules/no-unused-class
import styles from './Sidebar.module.scss'

const Sidebar: React.FC = () => {
    const [isOpened, setIsOpened] = useState(false)
    const [sidebarHeight, setSidebarHeight] = useState(0)
    const [sidebarItemsWidth, setSidebarItemsWidth] = useState<number[]>([])
    let sidebarItemsMaxWidth = 24

    if (sidebarItemsWidth.length) {
        sidebarItemsMaxWidth = Math.max(...sidebarItemsWidth)
    }

    const sidebarWidths = { min: 64, max: `${sidebarItemsMaxWidth + 74}px` }

    const sidebarRef = useCallback((node: HTMLDivElement) => {
        if (!node) { return }

        setSidebarHeight(node.getBoundingClientRect().height)
        const sidebarItems = node.querySelectorAll(`.${styles.prompt}`)
        const widths: number[] = []

        sidebarItems.forEach((item: Element) => {
            widths.push(item.getBoundingClientRect().width)
        })
        setSidebarItemsWidth(widths)
    }, [])

    const sidebarStyles = {
        top: `${window.innerHeight / 2 - sidebarHeight / 2}px`,
        width: isOpened ? sidebarWidths.max : sidebarWidths.min,
    }

    const dispatch = useDispatch()
    const handleClickToProfile = () => {
        dispatch(setModalIsActive(true))
    }

    const links = useSelector((state) => state.app.links)

    const userFirstName = useSelector((state) => state.profile.firstName)
    const userSecondName = useSelector((state) => state.profile.secondName)
    const userInitials = nameParser(userFirstName, userSecondName)

    return (
        <div
            className={clsx(
                styles.sidebar,
                'component',
                { [styles.opened]: isOpened },
            )}
            role="navigation"
            onDoubleClick={() => setIsOpened(!isOpened)}
            ref={sidebarRef}
            style={sidebarStyles}
        >
            <SidebarItem
                title="Profile"
                Ico={userInitials}
                isLinked={false}
                isSidebarOpened={isOpened}
                onClick={handleClickToProfile}
            />
            <SidebarItem
                title="Home"
                Ico={homeSvg}
                link={links.home}
                isSidebarOpened={isOpened}
            />
            <SidebarItem
                title="Accountancy"
                Ico={accountancySvg}
                link={links.accountancy}
                isSidebarOpened={isOpened}
            />
            <SidebarItem
                title="Statistics"
                Ico={statisticsSvg}
                link={links.statistics}
                isSidebarOpened={isOpened}
            />
            <SidebarItem
                title="Settings"
                Ico={settingsSvg}
                link={links.settings}
                isSidebarOpened={isOpened}
            />
            <SidebarItem
                title="Add&nbsp;operation"
                Ico={addSvg}
                isLinked={false}
                isSidebarOpened={isOpened}
            />
            <SidebarItem
                className={styles.expand}
                title="Expand&nbsp;sidebar"
                Ico={expandSvg}
                isSidebarOpened={isOpened}
                isLinked={false}
                onClick={() => setIsOpened(!isOpened)}
            />
        </div>
    )
}

export default Sidebar
