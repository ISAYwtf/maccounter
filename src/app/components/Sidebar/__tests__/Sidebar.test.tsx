import React from 'react'
import { compareSnapshot } from '@testUtils/renderSnapshot'
import Sidebar from '@components/Sidebar/Sidebar'
import { fireEvent, render, screen } from '@testing-library/react'

describe('Sidebar', () => {
    const renderComponent = <Sidebar />

    beforeEach(() => jest.useFakeTimers())

    // eslint-disable-next-line
    it('render Sidebar', () => compareSnapshot(renderComponent))

    it('dbClick should toggle data-open', () => {
        const { getByRole } = render(renderComponent)
        const sidebar = getByRole('navigation')
        let dataOpen

        fireEvent.dblClick(sidebar)
        dataOpen = sidebar.dataset.open
        expect(dataOpen).toStrictEqual('true')

        fireEvent.dblClick(sidebar)
        dataOpen = sidebar.dataset.open
        expect(dataOpen).toStrictEqual('false')
    })

    const checkEvent = (item: HTMLElement) => {
        fireEvent.mouseEnter(item)
        expect(setTimeout).toHaveBeenCalledTimes(1)
        expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 500)
        fireEvent.mouseLeave(item)

        jest.useFakeTimers()

        fireEvent.mouseEnter(item.children[0])
        expect(setTimeout).toHaveBeenCalledTimes(1)
        expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 500)
        fireEvent.mouseLeave(item.children[0])
    }

    // eslint-disable-next-line jest/expect-expect
    it('mouse enter should call timeout from sidebarItem profile', () => {
        render(renderComponent)

        const sidebarItemProfile = screen.getByTestId('sidebarItem-profile')
        checkEvent(sidebarItemProfile)
    })

    // eslint-disable-next-line jest/expect-expect
    it('mouse enter should call timeout from sidebarItems', () => {
        render(renderComponent)

        const sidebarItem = screen.getAllByRole('listitem')[1]
        checkEvent(sidebarItem)
    })
})
