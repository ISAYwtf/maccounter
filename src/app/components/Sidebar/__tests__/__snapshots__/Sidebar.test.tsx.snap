// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar render Sidebar 1`] = `
<DocumentFragment>
  <div
    class="sidebar component"
    data-open="false"
    role="navigation"
    style="top: 384px;"
  >
    <div
      class="sidebarItem"
      data-testid="sidebarItem-profile"
      role="listitem"
    >
      <span
        class="sidebarProfile"
      >
        IA
      </span>
      <div
        class="prompt"
      >
        <p>
          Profile
        </p>
      </div>
    </div>
    <div
      class="sidebarItem"
      role="listitem"
    >
      <svg>
        home.svg
      </svg>
      <div
        class="prompt"
      >
        <p>
          Home
        </p>
      </div>
    </div>
    <div
      class="sidebarItem"
      role="listitem"
    >
      <svg>
        accountancy.svg
      </svg>
      <div
        class="prompt"
      >
        <p>
          Accountancy
        </p>
      </div>
    </div>
    <div
      class="sidebarItem"
      role="listitem"
    >
      <svg>
        statistics.svg
      </svg>
      <div
        class="prompt"
      >
        <p>
          Statistics
        </p>
      </div>
    </div>
    <div
      class="sidebarItem"
      role="listitem"
    >
      <svg>
        settings.svg
      </svg>
      <div
        class="prompt"
      >
        <p>
          Settings
        </p>
      </div>
    </div>
    <div
      class="sidebarItem"
      role="listitem"
    >
      <svg>
        add.svg
      </svg>
      <div
        class="prompt"
      >
        <p>
          AddÂ operation
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
